admin-dev/themes/new-theme/js/components/grid/extension/preview-extension.js

```diff
@@ -0,0 +1,196 @@
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+const $ = window.$;
+
+/**
+ * Extends grid with preview functionality.
+ */
+export default class PreviewExtension {
+  constructor(previewRenderer) {
+    this.lock = [];
+    this.renderer = previewRenderer;
+    this.expandSelector = '.js-expand';
+    this.collapseSelector = '.js-collapse';
+    this.previewOpenClass = 'preview-open';
+    this.previewToggleSelector = '.preview-toggle';
+  }
+
+  /**
+   * Extends provided grid with preview functionality
+   *
+   * @param grid
+   */
+  extend(grid) {
+    const $gridContainer = $(grid.getContainer);
+
+    $gridContainer.find('tbody tr').on('mouseover mouseleave', event => this._handleIconHovering(event));
+    $gridContainer.find(this.previewToggleSelector).on('click', event => this._togglePreview(event));
+  }
+
+  /**
+   * Shows/hides preview toggling icons
+   *
+   * @param event
+   * @private
+   */
+  _handleIconHovering(event) {
+    const $previewToggle = $(event.currentTarget).find(this.previewToggleSelector);
+
+    if (event.type === 'mouseover' && !$(event.currentTarget).hasClass(this.previewOpenClass)) {
+      this._showExpandIcon($previewToggle);
+    } else {
+      this._hideExpandIcon($previewToggle);
+    }
+  }
+
+  /**
+   * Shows/hides preview
+   *
+   * @param event
+   * @private
+   */
+  _togglePreview(event) {
+    const $previewToggle = $(event.currentTarget);
+    const $columnRow = $previewToggle.closest('tr');
+
+    if ($columnRow.hasClass(this.previewOpenClass)) {
+      $columnRow.next('.preview-row').remove();
+      $columnRow.removeClass(this.previewOpenClass);
+      this._showExpandIcon($columnRow);
+      this._hideCollapseIcon($columnRow);
+
+      return;
+    }
+    const dataUrl = $(event.currentTarget).data('preview-data-url');
+
+    if (this._isLocked(dataUrl)) {
+      return;
+    }
+
+    // Prevents loading preview multiple times.
+    // Uses "dataUrl" as lock key.
+    this._lock(dataUrl);
+
+    $.ajax({
+      url: dataUrl,
+      method: 'GET',
+      dataType: 'json',
+    }).then((response) => {
+      this._renderPreviewContent($columnRow, response.preview);
+
+      this._unlock(dataUrl);
```

Comment:
Move this line in `complete` to don't duplicate it in catch method.

Reply:
I don't think it's a big deal, I wanted to use `Promise` API, but I can use jQuery's `complete` as well.

Reply:
If you want to use Promise, use `finally` :wink:

Reply:
I'll use old school `complete` since jQuery does not support `finally()` :man_shrugging: